### ISSUE TITLE
HGCN triplet loss lacks adaptive margins and hard-negative mining

### ISSUE THREAD

**Issue by lowmason** [lowmason/naics-embedder#62]
## Summary

The HGCN triplet loss uses a fixed margin and treats all sampled negatives equally. This is inconsistent with the more sophisticated Stage 3 contrastive loss and may slow convergence.

## Current Behavior
```python
# Fixed margin for all triplets
loss = max(0, margin + d(anchor, positive) - d(anchor, negative))
```

- Margin is static regardless of sample difficulty or embedding norms
- All negatives contribute equally to the loss
- Many negatives are "easy" (distant nodes) and provide little gradient signal

### 1. Adaptive Margins

Port the `NormAdaptiveMargin` approach from Stage 3's contrastive loss:
```python
# Margin adapts based on embedding norms or sample difficulty
adaptive_margin = base_margin * margin_scale_fn(anchor_norm, difficulty)
loss = max(0, adaptive_margin + d(anchor, positive) - d(anchor, negative))
```

### 2. Hard-Negative Mining

Prioritize negatives that are close to the anchor or violate the margin:
```python
# Option A: mine hardest negatives per batch
hard_negatives = select_hard_negatives(anchor, all_negatives, k=num_hard)

# Option B: reweight by difficulty
weights = compute_negative_weights(distances, temperature)
loss = weighted_triplet_loss(anchor, positive, negatives, weights)
```

Given the graph structure, many nodes are distant and trivially satisfy the margin—mining hard negatives should accelerate convergence.

## References

- Stage 3 uses `NormAdaptiveMargin` for text contrastive learning
- Hard-negative mining is standard practice in metric learning

## Priority

Medium—the report explicitly notes fixed margins as sub-optimal.
